### PR TITLE
first implementation of the triangular approximation tsp solution

### PR DESCRIPTION
**Pull request by tomasvicente3**

### CODE REVIEW

File: src/graph.cpp [hunk 1]
```diff
@@ -145,15 +170,36 @@ void Graph::kruskal(){
...
     }
-    return;
+}
+
+/**
+ * Calculates an approximation of the TSP, using the triangular approximation heuristic
+ * Time Complexity: //TODO
+ */
+void Graph::triangularTSPTour(){
+    /*
```
Comment (MRita443) [inline src/graph.cpp:194]: If the preorder tour uses the visited attribute, you should set all vertices to false here, or perhaps in kruskal as you are already iterating

File: src/graph.cpp [hunk 2]
```diff
@@ -145,15 +170,36 @@ void Graph::kruskal(){
...
+    -Build MST
+    -Get pre-order of the mst as vector //!DFS?
+    -Iterate that order getting total dist
+    */
+    kruskal();
+    tour = {0,{}};
+    preorderMSTTraversal(*(vertexSet.begin()));
+    return; //results are updated in tour
```
Comment (MRita443) [inline src/graph.cpp:202]: We are missing here adding the edge between the last vertex back to the first one

File: src/graph.cpp (hunk 1 above)
Comment (tomasvicente3) [reply src/graph.cpp:194]: Done!

File: src/graph.cpp [hunk 3]
```diff
@@ -145,15 +170,36 @@ void Graph::kruskal(){
  * Time Complexity: O(|V|+|E|)
```
[the remark this answers is not in the document]
Comment (tomasvicente3) [reply src/graph.cpp:170]: The struggle to find what you were refering to ahaha
Fixed

File: src/graph.cpp [hunk 4]
```diff
@@ -160,26 +155,42 @@ void Graph::kruskal(){
...
+    }
+    tour.course.push_back(stop);
+}
+
+/**
+ * Adds the distance between the first and last nodes so the tour "closes"
+ * Time Complexity: O(|E|) (average case) | O(|V|*|E|) (worst case)
+ */
+void Graph::tieDownTour(){
+    tour.distance += findEdge((*tour.course.begin())->getId(), (*tour.course.rbegin())->getId())->getLength();
```
Comment (MRita443) [inline src/graph.cpp:182]: I think not only should you add here the distance, but also the edge to the tour course, it is part of it afterall

File: src/graph.cpp [hunk 5]
```diff
@@ -132,10 +127,10 @@ void Graph::visitedDFS(const std::shared_ptr<Vertex> &source) {
  * Time Complexity: O(|E|log|E|)
  */
 void Graph::kruskal(){
-    std::list<Edge *> edges; //TODO: better generate this list (edgeSet like vertexSet?)
-    for (Vertex *v : vertexSet){
-        //v->setVisited(false); //TODO: Check if this is necessary
-        for (const std::shared_ptr<Edge>& e: v->getAdj()) {
+    std::list<std::shared_ptr<Edge>> edges; //TODO: better generate this list (edgeSet like vertexSet?)
```
Comment (MRita443) [inline src/graph.cpp:130]: We need a function here to add the edges that do not exist in the graph, with their distance, and possibly remove them later. Let me know what you think the best way to do this is

File: src/graph.cpp [hunk 6]
```diff
@@ -160,26 +155,42 @@ void Graph::kruskal(){
...
-void Graph::preorderMSTTraversal(Vertex *source){
+void Graph::preorderMSTTraversal(std::shared_ptr<Vertex> source){
     source->setVisited(true);
-    if (tour.second.size() != 0){
-        tour.first += findEdge((*(tour.second).rend())->getId(), source->getId())->getDist();
-    }
-    tour.second.push_back(source);
-    for (Edge *e : source->getAdj()){
+    addToTour(source);
+    for (std::shared_ptr<Edge> &e : source->getAdj()){
```
Comment (MRita443) [inline src/graph.cpp:193]: We're gonna have to change this here because not only should we consider the existing edges (ones in adj) but also the ones that do no exist

File: src/graph.h [hunk 7]
```diff
@@ -8,55 +8,65 @@
...
 
 class Graph {
   private:
     unsigned int totalEdges = 0;
-    std::vector<Vertex *> vertexSet;    // vertex set
-    std::unordered_map<unsigned int, Vertex *> idToVertex;
-    std::pair<unsigned int, std::list<Vertex *>> tour;
+//?    std::unordered_map<unsigned int, std::shared_ptr<Vertex>> idToVertex;
+    struct tour_t {
+        unsigned int distance;
```
Comment (MRita443) [inline src/graph.h:21]: This should be a long, as an Edge's length is a double

**Review by MRita443 — CHANGES_REQUESTED** [review]

File: src/graph.cpp (hunk 4 above)
Comment (tomasvicente3) [reply src/graph.cpp:182]: Fixed

File: src/graph.h (hunk 7 above)
Comment (tomasvicente3) [reply src/graph.h:21]: It should be a double, shouldn't it?

File: src/graph.cpp (hunk 5 above)
Comment (tomasvicente3) [reply src/graph.cpp:130]: ### Taken care in call

File: src/graph.cpp (hunk 6 above)
Comment (tomasvicente3) [reply src/graph.cpp:193]: ### Taken care in call

File: src/coordinates.cpp [hunk 8]
```diff
@@ -6,6 +6,7 @@ double Coordinates::distanceTo(Coordinates point) const {
     double lat1 = this->latitude, lat2 = point.latitude;
     double lon1 = this->longitude, lon2 = point.longitude;
 
+    if ((lat1 == 0 && lon1 == 0) || (lat2 == 0 && lon2 == 0)) return -1;
```
Comment (MRita443) [inline src/coordinates.cpp:9]: Why is this needed?

File: src/graph.cpp [hunk 9]
```diff
@@ -1,5 +1,5 @@
+#include <set>
```
Comment (MRita443) [inline src/graph.cpp:1]: If you need this you should place it in the .h file

File: src/graph.cpp [hunk 10]
```diff
@@ -64,59 +81,141 @@ Graph::addBidirectionalEdge(const unsigned int &source, const unsigned int &dest
...
+            printTour();
+            break;
+        case -1:
+            printf("Couldn't calculate approximation of TSP for this graph!\n");
+            break;
+        case -2:
+            printf("Course would contain self-loop!\n");
+            break;
+    }
+    //preorderMSTTraversal(findVertex(0));
```
Comment (MRita443) [inline src/graph.cpp:202]: This is still needed, otherwise this is not the algorithm we are looking for:
image

We just switched the mst algorithm but we need a pre order walk nonetheless

**Review by MRita443 — CHANGES_REQUESTED** [review]

File: src/coordinates.cpp (hunk 8 above)
Comment (tomasvicente3) [reply src/coordinates.cpp:9]: So if there aren't coordinates set, it can identify and raise the appropriate error message (for the shipping graph)

File: src/graph.cpp (hunk 9 above)
Comment (tomasvicente3) [reply src/graph.cpp:1]: Oof true

File: src/graph.cpp (hunk 10 above)
Comment (tomasvicente3) [reply src/graph.cpp]: I thought it was already implemented midst Prim's execution, however, it wasn't a dfs nor a optimal traversal

File: src/coordinates.cpp (hunk 8 above)
Comment (MRita443) [reply src/coordinates.cpp:9]: What if the coordinates are set and they are set to 0? Wouldn't that influence the other graphs?

File: src/graph.cpp [hunk 11]
```diff
@@ -162,18 +166,24 @@ int Graph::addToTour(std::shared_ptr<Vertex> stop) {
...
     }
 
-    if (source->getId() == 0) addToTour(source);
-*/}
+    if (source->getId() == 0) return addToTour(source);
+    return 0;
+}
 
 /**
  * Calculates an approximation of the TSP, using the triangular approximation heuristic
```
Comment (MRita443) [inline src/graph.cpp:189]: This complexity here in theory is O(V + E), I am not sure in our case it would be O(V + E + E*logV) so O(V + E(logV + 1)) so perhaps just O(V + E*logV), I don't know, let me know what you think

**Review by MRita443 — CHANGES_REQUESTED** [review]

File: src/graph.cpp [hunk 12]
```diff
@@ -162,18 +166,24 @@ int Graph::addToTour(std::shared_ptr<Vertex> stop) {
...
+        std::shared_ptr<Vertex> dest = findVertex(i);
+
+        if(!(dest->isVisited()) && selectedEdges[source->getId()][i]){
+            preorderMSTTraversal(dest);
         }
     }
 
-    if (source->getId() == 0) addToTour(source);
-*/}
+    if (source->getId() == 0) return addToTour(source);
```
[the remark this answers is not in the document]
Comment (tomasvicente3) [reply src/graph.cpp:184]: No, because its recursive, and by the time this function ends, all the "children" functions have already ended. So the node 0 only appears at the start and end of the tour

[the remark this answers is not in the document]
Comment (MRita443) [reply src/graph.cpp:184]: Ok I get it makes sense, but I think the if check is a bit unnecessary then, why only do it when the source is 0?

File: src/coordinates.cpp (hunk 8 above)
Comment (tomasvicente3) [reply src/coordinates.cpp:9]: That's on you, that set the default to (0,0) 😭 But It's a huge coincidence if exists a point exactly on (0,0)

File: src/graph.cpp [hunk 13]
```diff
@@ -97,10 +97,12 @@ void Graph::visitedDFS(const std::shared_ptr<Vertex> &source) {
 /**
  * @brief Builds a MST using Prim's algorithm
  * Time Complexity: O(|E|log|E|)
```
[the remark this answers is not in the document]
Comment (tomasvicente3) [reply src/graph.cpp]: It's O(V^2)

File: src/graph.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (tomasvicente3) [reply src/graph.cpp:183]: To close out the tour, so the first node in the tour will be at the end too (I can move it outside the function, but I think its cleaner this way (Even though it will run this check for every other node))

File: src/graph.cpp (hunk 11 above)
Comment (tomasvicente3) [reply src/graph.cpp:188]: It's O(V^2) as this function only calls other functions and the max complexity of the other functions is O(V^2)

File: src/coordinates.cpp (hunk 8 above)
Comment (MRita443) [reply src/coordinates.cpp:9]: Ok, I just checked there is not

File: src/graph.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (MRita443) [reply src/graph.cpp:183]: Oh I get it now, because you don't wanna close out the tour with other source ids when it finishes running

**Review by MRita443 — APPROVED** [review]
